clippy_lints/src/needless_question_mark.rs

```diff
@@ -0,0 +1,191 @@
+use rustc_errors::Applicability;
+use rustc_hir::def::{CtorKind, CtorOf, DefKind, Res};
+use rustc_hir::{Body, Expr, ExprKind, LangItem, MatchSource, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::DefIdTree;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+use crate::utils;
+use if_chain::if_chain;
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// Suggest for terminating expressions
+    /// ```rust,ignore
+    /// value.inner_option;
+    /// value.inner_result;
+    /// ```
+    /// instead of
+    /// ```rust,ignore
+    /// Some(value.inner_option?);
+    /// Ok(value.inner_result?);
+    /// ```
+    ///
+    /// **Why is this bad?** There's no reason to use ? to short-circuit when execution of the body will end there anyway.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // struct Outer {
+    /// //     inner: Option<String>
+    /// // }
+    /// //
+    /// // let a = Some(Outer { inner: Some("hello".into()) });
+    /// // a.and_then(|a| Some(a.inner?))
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // struct Outer {
+    /// //     inner: Option<String>
+    /// // }
+    /// //
+    /// // let a = Some(Outer { inner: Some("hello".into()) });
+    /// // a.and_then(|a| a.inner)
+    /// ```
+    pub NEEDLESS_QUESTION_MARK,
+    complexity,
+    "Suggest value.inner_option instead of Some(value.inner_option?). The same goes for Result<T, E>."
+}
+
+declare_lint_pass!(NeedlessQuestionMark => [NEEDLESS_QUESTION_MARK]);
+
+#[derive(Debug)]
+enum SomeOkCall<'a> {
+    SomeCall(&'a Expr<'a>, &'a Expr<'a>),
+    OkCall(&'a Expr<'a>, &'a Expr<'a>),
+}
+
+impl LateLintPass<'_> for NeedlessQuestionMark {
+    /*
+     * The question mark operator is compatible with both Result<T, E> and Option<T>,
+     * from Rust 1.13 and 1.22 respectively.
```

Comment:
Please add the `msrv` configuration for this lint then. See the adding_lints.md documentation on how to do this.

Reply:
Would it be advisable to declare two MSRVs for this lint? One for Option, one for Result

Reply:
This lint should appear only once in `conf.rs`, but I would use two constants (one for option and one for result) here and check them separately, yes.